legion/engine/core/async/job_pool.hpp

```diff
@@ -59,9 +75,15 @@ namespace legion::core::async
                 {
                     auto* job = jobPoolPtr->pop_job();
                     if (job)
+                    {
                         job->execute();
-                    else
-                        this->m_progress->complete();
+                        jobPoolPtr->complete_job();
+                    }
+
+                    if (jobPoolPtr->is_done())
+                    {
+                        m_onComplete();
+                    }
```

Comment:
This code is repeated at line 95, perhpaps a function like executeAndCompleteJob(Job job) can be made